### PR TITLE
	Improve performance of REST API

### PR DESCRIPTION
**Pull request by Kamidu**

### CODE REVIEW

File: modules/mobile-agents/android/jax-rs/src/main/java/org/wso2/carbon/mdm/services/android/util/AndroidAPIUtils.java [hunk 1]
```diff
@@ -114,8 +114,8 @@ public static Response getOperationResponse(List<String> deviceIDs, Operation op
         if (status > 0) {
             GCMService gcmService = getGCMService();
             if (gcmService.isGCMEnabled()) {
-                List<Device> devices = new ArrayList<Device>(validDeviceIds.size());
-                for (DeviceIdentifier deviceIdentifier : validDeviceIds) {
+                List<Device> devices = new ArrayList<Device>(deviceIDHolder.getValidDeviceIDList().size());
+                for (DeviceIdentifier deviceIdentifier : deviceIDHolder.getValidDeviceIDList()) {
```
Comment (harshanL) [inline modules/mobile-agents/android/jax-rs/src/main/java/org/wso2/carbon/mdm/services/android/util/AndroidAPIUtils.java]: Assign deviceIDHolder.getValidDeviceIDList() to a variable and use it here.

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java [hunk 2]
```diff
@@ -421,8 +422,12 @@ public Response getAllUsers() throws MDMAPIException {
         }
         ResponsePayload responsePayload = new ResponsePayload();
         responsePayload.setStatusCode(HttpStatus.SC_OK);
+        int count = 0;
+        if(userList != null){
```
Comment (harshanL) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: Fix formatting of if block above.

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java [hunk 3]
```diff
@@ -460,8 +466,12 @@ public Response getMatchingUsers(@PathParam ("filter") String filter) throws MDM
...
```
Comment (harshanL) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: Fix formatting of if block above.

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java [hunk 4]
```diff
@@ -498,8 +509,12 @@ public Response getAllUsersByUsername(@QueryParam("username") String userName) t
...
```
Comment (harshanL) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: Fix formatting of if block above.

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java [hunk 5]
```diff
@@ -531,8 +547,12 @@ public Response getAllUserNamesByUsername(@QueryParam("username") String userNam
...
```
Comment (harshanL) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: Fix formatting of if block above.

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/util/MDMUtil.java [hunk 6]
```diff
@@ -32,7 +32,7 @@ public static Profile convertProfile(org.wso2.carbon.mdm.beans.Profile mdmProfil
         profile.setDeviceType(mdmProfile.getDeviceType());
 
         List<org.wso2.carbon.policy.mgt.common.ProfileFeature> profileFeatures = new ArrayList<org.wso2.carbon.policy
-                .mgt.common.ProfileFeature>();
+                .mgt.common.ProfileFeature>(mdmProfile.getProfileFeaturesList().size());
         for(ProfileFeature mdmProfileFeature:mdmProfile.getProfileFeaturesList()){
```
Comment (harshanL) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/util/MDMUtil.java]: Fix formatting of for loop above.

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/Policy.java [hunk 7]
```diff
@@ -232,7 +232,7 @@ public Response updatePolicyPriorities(List<PriorityUpdatedPolicyWrapper> priori
             throws MDMAPIException {
         PolicyManagerService policyManagementService = MDMAPIUtils.getPolicyManagementService();
         List<org.wso2.carbon.policy.mgt.common.Policy> policiesToUpdate =
-                new ArrayList<org.wso2.carbon.policy.mgt.common.Policy>();
+                new ArrayList<org.wso2.carbon.policy.mgt.common.Policy>(priorityUpdatedPolicies.size());
```
Comment (kasundananjaya) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/Policy.java]: Fix formatting issues.

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java [hunk 8]
```diff
@@ -402,9 +402,10 @@ public Response getAllUsers() throws MDMAPIException {
             log.debug("Getting the list of users with all user-related information");
         }
         UserStoreManager userStoreManager = MDMAPIUtils.getUserStoreManager();
-        ArrayList<UserWrapper> userList = new ArrayList<UserWrapper>();
+        ArrayList<UserWrapper> userList;
         try {
             String[] users = userStoreManager.listUsers("*", -1);
+            userList = new ArrayList<UserWrapper>(users.length);
```
Comment (kasundananjaya) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: Fix formatting issues.

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java [hunk 9]
```diff
@@ -441,9 +446,10 @@ public Response getMatchingUsers(@PathParam ("filter") String filter) throws MDM
             log.debug("Getting the list of users with all user-related information using the filter : " + filter);
         }
         UserStoreManager userStoreManager = MDMAPIUtils.getUserStoreManager();
-        ArrayList<UserWrapper> userList = new ArrayList<UserWrapper>();
+        ArrayList<UserWrapper> userList;
         try {
             String[] users = userStoreManager.listUsers( filter + "*", -1);
+            userList = new ArrayList<UserWrapper>(users.length);
```
Comment (kasundananjaya) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: Fix formatting issues.

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java (hunk 2 above)
Comment (Kamidu) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: These formatting issues has been fixed in a recent PR

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java (hunk 3 above)
Comment (Kamidu) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: These formatting issues has been fixed in a recent PR

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java (hunk 4 above)
Comment (Kamidu) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: These formatting issues has been fixed in a recent PR

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java (hunk 5 above)
Comment (Kamidu) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: These formatting issues has been fixed in a recent PR

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/util/MDMUtil.java (hunk 6 above)
Comment (Kamidu) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/util/MDMUtil.java]: Recommended fixes  can be found in PR[1]

[1]. https://github.com/wso2/product-mdm/pull/158

File: modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java [hunk 10]
```diff
@@ -402,9 +402,10 @@ public Response getAllUsers() throws MDMAPIException {
             log.debug("Getting the list of users with all user-related information");
         }
         UserStoreManager userStoreManager = MDMAPIUtils.getUserStoreManager();
-        ArrayList<UserWrapper> userList = new ArrayList<UserWrapper>();
+        ArrayList<UserWrapper> userList;
```
Comment (prabathabey) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: As a best practice, always bind variable types to their respective abstractions for better extensibility of the code.

Comment (Kamidu) [inline modules/apps/jax-rs/mdm-admin/src/main/java/org/wso2/carbon/mdm/api/User.java]: Recommended fixes can be found in PR[1]

[1]. https://github.com/wso2/product-mdm/pull/166

**Pull request merged**